### PR TITLE
Add metadata to upload

### LINKED ISSUE #235: Post activity with attachment should upload metadata
## Background

We recently updated our `/upload` endpoint to support metadata (e.g. `thumbnailUrl`) for attachments. The service work is carried out by @NickEricson and did not introduce backward compatibility issues.

This work is required for https://github.com/microsoft/BotFramework-WebChat/issues/2422.

## Changes

Instead of removing `activity.attachments`, we leave the `attachments` array but only removing `activity.attachments[].contentUrl`. This will send `contentType` and `thumbnailUrl` to the service.

The service will match metadata by looking at:
- `activity.attachments[].name`
- `filename` of multipart, where the multipart has `name` equals to `file`

#### Today

```
-------boundary
Content-Disposition: form-data; name="activity"; filename="blob"
Content-Type: application/vnd.microsoft.activity

{
  "from": {
    "id": "dl_1234567"
  },
  "type": "message",
  ...
}
-------boundary
Content-Disposition: form-data; name="file"; filename="abc.png"
Content-Type: image/png

...
-------boundary
```

#### Tomorrow

```diff
  -------boundary
  Content-Disposition: form-data; name="activity"; filename="blob"
  Content-Type: application/vnd.microsoft.activity

  {
+   "attachments": [{
+     "contentType": "image/png",
+     "name": "abc.png",
+     "thumbnailUrl": "data:image/jpeg;base64,..."
+   }],
    "from": {
      "id": "dl_1234567"
    },
    "type": "message",
    ...
  }
  -------boundary
  Content-Disposition: form-data; name="file"; filename="abc.png"
  Content-Type: image/png

  ...
  -------boundary
```

> Note: the `filename` field on multipart is used for matching the corresponding item in the `attachments` array.

### PR DESCRIPTION
**Pull request by compulim**
> Fix #235.

## Background

We recently updated our `/upload` endpoint to support metadata (e.g. `thumbnailUrl`) for attachments. The service work is carried out by @NickEricson and did not introduce backward compatibility issues.

This work is required for https://github.com/microsoft/BotFramework-WebChat/issues/2422.

## Changes

Instead of removing `activity.attachments`, we leave the `attachments` array but only removing `activity.attachments[].contentUrl`. This will send `contentType` and `thumbnailUrl` to the service.

The service will match metadata by looking at:
- `activity.attachments[].name`
- `filename` of multipart, where the multipart has `name` equals to `file`

#### Today

```
-------boundary
Content-Disposition: form-data; name="activity"; filename="blob"
Content-Type: application/vnd.microsoft.activity

{
  "from": {
    "id": "dl_1234567"
  },
  "type": "message",
  ...
}
-------boundary
Content-Disposition: form-data; name="file"; filename="abc.png"
Content-Type: image/png

...
-------boundary
```

#### Tomorrow

```diff
  -------boundary
  Content-Disposition: form-data; name="activity"; filename="blob"
  Content-Type: application/vnd.microsoft.activity

  {
+   "attachments": [{
+     "contentType": "image/png",
+     "name": "abc.png",
+     "thumbnailUrl": "data:image/jpeg;base64,..."
+   }],
    "from": {
      "id": "dl_1234567"
    },
    "type": "message",
    ...
  }
  -------boundary
  Content-Disposition: form-data; name="file"; filename="abc.png"
  Content-Type: image/png

  ...
  -------boundary
```

> Note: the `filename` field on multipart is used for matching the corresponding item in the `attachments` array.

### Added

- Fix [#235](https://github.com/Microsoft/BotFramework-DirectLineJS/issues/235). Fix Added metadata when uploading attachments, including `thumbnailUrl`, by [@compulim](https://github.com/compulim), in PR [#236](https://github.com/Microsoft/BotFramework-DirectLineJS/pull/236)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -61,6 +61,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
    - [`webpack@4.39.3`](https://npmjs.com/package/webpack)
    - [`webpack-cli@3.3.8`](https://npmjs.com/package/webpack-cli)
 
+### Added
+- Fix [#235](https://github.com/Microsoft/BotFramework-DirectLineJS/issues/235). Fix Added metadata when uploading attachments, including `thumbnailUrl`, by [@compulim](https://github.com/compulim), in PR [#236](https://github.com/Microsoft/BotFramework-DirectLineJS/pull/236)
```
Comment (tdurnford) [inline CHANGELOG.md]:
```suggestion
- Fix [#235](https://github.com/Microsoft/BotFramework-DirectLineJS/issues/235). Added metadata when uploading attachments, including `thumbnailUrl`, by [@compulim](https://github.com/compulim), in PR [#236](https://github.com/Microsoft/BotFramework-DirectLineJS/pull/236)
```

File: src/dedupeFilenames.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import parseFilename from './parseFilename';
+
+export default function dedupeFilenames(array: string[]) {
+    const nextArray: string[] = [];
+
+    array.forEach(value => {
+        let count = 0;
+        let nextValue = value;
+		const { extname, name } = parseFilename(value);
```
Comment (tdurnford) [inline src/dedupeFilenames.ts]: Remove extra tabs

Comment (compulim) [reply src/dedupeFilenames.ts]: 🙈

image

**Comment by compulim** [comment]
Test added and passing. Phew~~~

**Comment by NickEricson** [comment]
Thanks @compulim . Look forward to seeing this merged.

File: __tests__/happy.uploadAttachments.js [hunk 3]
```diff
@@ -87,17 +89,25 @@ describe('Happy path', () => {
...
+            );
 
             // Use the commented code below after bug #194 is fixed.
             // https://github.com/microsoft/BotFramework-DirectLineJS/issues/194
 
-            // await expect(fetchAsBase64(attachments[0].contentUrl)).resolves.toBe(await fetchAsBase64(activityFromUser.attachments[0].contentUrl));
-            // await expect(fetchAsBase64(attachments[1].contentUrl)).resolves.toBe(await fetchAsBase64(activityFromUser.attachments[1].contentUrl));
-
-            return true;
+            // return (
```
Comment (tdurnford) [inline __tests__/happy.uploadAttachments.js]: Do we want to keep this commented out?

Comment (compulim) [reply __tests__/happy.uploadAttachments.js]: Yap, keep commenting out so we don't need to think how to rewrite that code, later.

**Pull request merged**